src/grain/distribution.js

```diff
@@ -0,0 +1,232 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating distribution amounts for
+ * contributors. A distribution contains "receipts" showing how much Grain each
+ * contributor will receive, the timestamp of the distribution in question, and the
+ * "strategy" used to distribute grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which distributes a fixed amount of grain based on the cred scores
+ * in the most recent completed time interval
+ * - LIFETIME, which distributes a fixed amount of grain based on cred scores
+ * across
+ * all time, prioritizing paying people who were under-paid historically (i.e.
+ * their lifetime earnings are lower than we would expect given their current
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {type Grain, multiplyFloat, toFloatRatio, ZERO} from "./grain";
+
+export const DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | LifetimeV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type LifetimeV1 = {|
+  +type: "LIFETIME",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type DistributionV1 = {|
+  +type: "DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Distribution given:
+ * - the strategy we're using
+ * - the full cred history for all users
+ * - the lifetime earnings of all users
+ * - the timestamp for the distribution
+ */
+export function distribution(
+  strategy: DistributionStrategy,
+  credHistory: CredHistory,
+  earnings: Map<NodeAddressT, Grain>,
+  timestampMs: number
+): DistributionV1 {
+  const filteredSlices = credHistory.filter(
+    (s) => s.intervalEndMs <= timestampMs
+  );
```

Comment:
The reason I wanted the timestampMs is to support the use case where you want to do "retroactive" issuance, e.g. "the project is one year old, we want a policy of doing 5k immediate / 10k lifetime every week, let's simulate history and pretend we did 52 immediate distributions at 5k each, and then one lifetime payment of 520k". not clear if this is something ppl will want to do in practice, but it seems like a valuable feature to allow for.

for this, we need to know the time-sliced cred so we can simulate the past distributions. however, we never need the earnings sliced by time, since the lifetime / balanced payment always uses the _full_ history of payouts, since you're iterating and adding new (historical) payments one-at-a-time, but you're never putting in payments out of order.

does that make sense? basically i just can't see any use case where we need timestamps associated with the earnings.

Reply:
> for this, we need to know the time-sliced cred so we can simulate the past distributions. however, we never need the earnings sliced by time, since the lifetime / balanced payment always uses the full history of payouts, since you're iterating and adding new (historical) payments one-at-a-time, but you're never putting in payments out of order.

There's an assumptions here that is outside the scope of this function: The reason you're calling this function is to append the resulting event to the same ledger you're given the aggregate of.

What if we had a weight change, and we want to create a comparison between "what if we had this weight change in place x weeks ago" and "the actual distribution history". You'll want to filter the ledger to be "what it looked like at that point" same as we're doing with cred here.

It's debatable which one of these use cases is more obscure. But if we want to allow cred timeline filtering to allow more use cases then it would make sense to do that for the distribution history too. Or we could consider it possibly obscure and push filtering responsibilities for both to the caller.

I'm leaning towards: generally speaking possible future use-cases aren't a great reason to add functionality. And in this case the caller could implement it as needed. So let's not filter for either and let the caller do that.

Reply:
You raise some reasonable points. However, it also falls in the category of "easy to fix later, whenever we need more/different functionality", since it's a local API that we don't currently plan to expose to external consumers. We can follow up as part of https://github.com/sourcecred/sourcecred/issues/1759